### PR TITLE
RATIS-1925. Support Zero-Copy in GrpcClientProtocolService

### PR DESCRIPTION
**Pull request by duongkame**
## What changes were proposed in this pull request?

Support Zero-Copy in GrpcClientProtocolService

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-1925

## How was this patch tested?

Added unit tests.

### CODE REVIEW

**Comment by duongkame** [comment]
@szetszwo please have a look.

**Review by szetszwo — COMMENTED** [review]
@duongkame , thanks a lot for working on this!  The change looks good.

Do we really need `zeroCopyEnabled`?  It seems that there are no reason disabling it.

**Comment by duongkame** [comment]
> Do we really need `zeroCopyEnabled`? It seems that there are no reason disabling it.

Zero-Copy may not be naturally compatible with all use cases. One edge case I found is that a StateMachine may cache LogEntry's data and use it outside the Ratis control. If zero-copy LogEntry data is cached and then used after Ratis releases the backing buffer, the JVM will crash with memory corruption.
The feature should be deliberately enabled per use case. Maybe we need to update `configurations.md` too.

**Comment by szetszwo** [comment]
> ... a StateMachine may cache LogEntry's data and use it outside the Ratis control. ...

We should pass the `ReferenceCountedObject` in this case.  Indeed, Ozone currently cache log entries in datanode. We should make it work.

**Comment by duongkame** [comment]
> > ... a StateMachine may cache LogEntry's data and use it outside the Ratis control. ...
> 
> We should pass the `ReferenceCountedObject` in this case. Indeed, Ozone currently cache log entries in datanode. We should make it work.

Yeah, it's where I'm from. Agree, I filed RATIS-1997 for this. Btw, for backward compatibility, we need to disable zero copy by default. When applications enable zero-copy, a migration to use the new versions of StateMachine API is needed.

**Comment by szetszwo** [comment]
For backward compatibility, we will provide two APIs.  For examples, there are two `StateMachine.DataChannel.write(..)` methods:
```java
//StateMachine.DataChannel
    default int write(ByteBuffer buffer) throws IOException {...}

    default int write(ReferenceCountedObject<ByteBuffer> referenceCountedBuffer) throws IOException {...}
```
An implementation can choose to override either one of them.

**Comment by duongkame** [comment]
> For backward compatibility, we will provide two APIs.  For examples, there are two `StateMachine.DataChannel.write(..)` methods:
> 
> ```java
> 
> //StateMachine.DataChannel
> 
>     default int write(ByteBuffer buffer) throws IOException {...}
> 
> 
> 
>     default int write(ReferenceCountedObject<ByteBuffer> referenceCountedBuffer) throws IOException {...}
> 
> ```
> 
> An implementation can choose to override either one of them.

Thanks @szetszwo, totally agree with the API suggestion. We will achive code compatibility with that. 
By backward compatibility, I meant for exsting applications, they have to make a code change to implement the new API before enabling zero copy safely or may crash in runtime. Releasing a version of Ratis with zero-copy implicitly enabled doesn't seem backward compatible in that sense. We need a flag that is off by default, or zero copy is released in a new major version with clear release notes stating the needed migration.

**Comment by szetszwo** [comment]
> ... for exsting applications, they have to make a code change to implement the new API before enabling zero copy safely or may crash in runtime.

We will have zero copy enabled and  copy log object before passing it in the old API (i.e. the method without `ReferenceCountedObject`).  For the new API with `ReferenceCountedObject`, we will not copy it.

**Comment by szetszwo** [comment]
The code looks like below:
```java
 default int write(LogEntryProto entry) throws IOException {
    throw new UnsupportedOperationException();
 }

 default int write(ReferenceCountedObject<LogEntryProto> referenceCountedEntry) throws IOException {
    LogEntryProto entry = copy(referenceCountedEntry.get())
    return write();
 }
```
Applications can override either of the methods.  The ratis code will call `write(ReferenceCountedObject<LogEntryProto>)`.

**Comment by duongkame** [comment]
> > ... for exsting applications, they have to make a code change to implement the new API before enabling zero copy safely or may crash in runtime.
> 
> We will have zero copy enabled and copy log object before passing it in the old API (i.e. the method without `ReferenceCountedObject`). For the new API with `ReferenceCountedObject`, we will not copy it.

Ok, that path makes sense. I removed the config flag.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by duongkame** [comment]
The `TestMessageStreamApiWithGrpc.testStream` seems to become flaky after this chance. The failure looks quite serious. 
The test looks like below:
```
   // stream multiple parts
    final int numParts = 9;
    final int endOfRequest = 6;
    final StringBuilder key = new StringBuilder();
    try(RaftClient client = cluster.createClient();
        MessageOutputStream out = client.getMessageStreamApi().stream()) {
      for (int i = 1; i <= numParts; i++) {
        key.append(i);
        SimpleMessage message = new SimpleMessage(i + "");
        boolean end = i == endOfRequest;
        LOG.info("Sending: {}, end = {}",  message.messageId, end);
        out.sendAsync(message, end);
      }
    }
    // check if all the parts are streamed as a single message.
```
I think two messages `123456` and `789` is supposed to be committed.
However, the logs of the failures show that other messages are committed, e.g.

```
2024-01-11 14:46:46,921 [Time-limited test] INFO  grpc.TestMessageStreamApiWithGrpc (MessageStreamApiTests.java:runTestStream(71)) - Sending: 1, end = false
2024-01-11 14:46:46,931 [Time-limited test] INFO  grpc.TestMessageStreamApiWithGrpc (MessageStreamApiTests.java:runTestStream(71)) - Sending: 2, end = false
2024-01-11 14:46:46,931 [Time-limited test] INFO  grpc.TestMessageStreamApiWithGrpc (MessageStreamApiTests.java:runTestStream(71)) - Sending: 3, end = false
2024-01-11 14:46:46,932 [Time-limited test] INFO  grpc.TestMessageStreamApiWithGrpc (MessageStreamApiTests.java:runTestStream(71)) - Sending: 4, end = false
2024-01-11 14:46:46,932 [Time-limited test] INFO  grpc.TestMessageStreamApiWithGrpc (MessageStreamApiTests.java:runTestStream(71)) - Sending: 5, end = false
2024-01-11 14:46:46,932 [Time-limited test] INFO  grpc.TestMessageStreamApiWithGrpc (MessageStreamApiTests.java:runTestStream(71)) - Sending: 6, end = true
2024-01-11 14:46:46,932 [Time-limited test] INFO  grpc.TestMessageStreamApiWithGrpc (MessageStreamApiTests.java:runTestStream(71)) - Sending: 7, end = false
2024-01-11 14:46:46,932 [Time-limited test] INFO  grpc.TestMessageStreamApiWithGrpc (MessageStreamApiTests.java:runTestStream(71)) - Sending: 8, end = false
2024-01-11 14:46:46,932 [Time-limited test] INFO  grpc.TestMessageStreamApiWithGrpc (MessageStreamApiTests.java:runTestStream(71)) - Sending: 9, end = false
....
2024-01-11 14:46:46,995 [s0@group-F60956FB3D4A-StateMachineUpdater] INFO  impl.SimpleStateMachine4Testing (SimpleStateMachine4Testing.java:put(205)) - s0: put 1, 623456 -> (t:1, i:1), STATEMACHINELOGENTRY, 46@client-B3C3E9191731
2024-01-11 14:46:46,998 [s0@group-F60956FB3D4A-StateMachineUpdater] INFO  impl.SimpleStateMachine4Testing (SimpleStateMachine4Testing.java:put(205)) - s0: put 2, 789 -> (t:1, i:2), STATEMACHINELOGENTRY, 50@client-B3C3E9191731
2024-01-11 14:46:47,012 [s2@group-F60956FB3D4A-StateMachineUpdater] INFO  impl.SimpleStateMachine4Testing (SimpleStateMachine4Testing.java:put(205)) - s2: put 1, 623456 -> (t:1, i:1), STATEMACHINELOGENTRY, 46@client-B3C3E9191731
2024-01-11 14:46:47,012 [s1@group-F60956FB3D4A-StateMachineUpdater] INFO  impl.SimpleStateMachine4Testing (SimpleStateMachine4Testing.java:put(205)) - s1: put 1, 623456 -> (t:1, i:1), STATEMACHINELOGENTRY, 46@client-B3C3E9191731
2024-01-11 14:46:47,023 [s2@group-F60956FB3D4A-StateMachineUpdater] INFO  impl.SimpleStateMachine4Testing (SimpleStateMachine4Testing.java:put(205)) - s2: put 2, 789 -> (t:1, i:2), STATEMACHINELOGENTRY, 50@client-B3C3E9191731
2024-01-11 14:46:47,023 [s1@group-F60956FB3D4A-StateMachineUpdater] INFO  impl.SimpleStateMachine4Testing (SimpleStateMachine4Testing.java:put(205)) - s1: put 2, 789 -> (t:1, i:2), STATEMACHINELOGENTRY, 50@client-B3C3E9191731
```
I'm debugging to understand the issue, but it'll be very helpful if you can have a quick look.  @szetszwo

**Comment by szetszwo** [comment]
@duongkame , the grpc tests keep failing.  This failure seems related to buffer corruption https://github.com/apache/ratis/actions/runs/7493450746/job/20404797679?pr=1007#step:6:656

**Comment by duongkame** [comment]
> @duongkame , the grpc tests keep failing. This failure seems related to buffer corruption https://github.com/apache/ratis/actions/runs/7493450746/job/20404797679?pr=1007#step:6:656

Turns out, it's specific to MessageStream API. `MessageStreamRequests.PendingStream` keeps a raw reference to the message buffer of all the pending request.

In `MessageStreamRequests.PendingStream.append`:
```
synchronized CompletableFuture<ByteString> append(long messageId, Message message) {
      ...
      bytes = bytes.concat(message.getContent());
      return CompletableFuture.completedFuture(bytes);
    }
```
The `bytes.concat(message.getContent())` only creates a wrapper of the two `ByteStrings` (RopeByteString). And this makes the resulting `bytes` keep references to all of the pending message request buffers without reference counting.

We can either just copy the message. 
```bytes = bytes.concat(copy(message.getContent()));```
or make PendingStream keeping all the counters to all the previous requests in the stream.

Think for this case we should just copy. @szetszwo

File: ratis-server/src/main/java/org/apache/ratis/server/impl/MessageStreamRequests.java [hunk 1]
```diff
@@ -39,25 +42,32 @@ private static class PendingStream {
...
+      ReferenceCountedObject<ByteString> bytesRef = ReferenceCountedObject.delegateFrom(pendingRefs, bytes);
+      return CompletableFuture.completedFuture(bytesRef);
     }
 
-    synchronized CompletableFuture<ByteString> getBytes(long messageId, Message message) {
-      return append(messageId, message);
+    synchronized CompletableFuture<ReferenceCountedObject<ByteString>> getBytes(long messageId,
+        ReferenceCountedObject<Message> messageRef) {
+      return append(messageId, messageRef);
     }
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/MessageStreamRequests.java:71]: Make it return `CompletableFuture<ReferenceCountedObject<RaftClientRequest>>`.  The code will become simpler
```java
    synchronized CompletableFuture<ReferenceCountedObject<RaftClientRequest>> getWriteRequest(long messageId,
        ReferenceCountedObject<RaftClientRequest> requestRef) {
      return append(messageId, requestRef)
          .thenApply(bytes -> RaftClientRequest.toWriteRequest(requestRef.get(), () -> bytes))
          .thenApply(write -> ReferenceCountedObject.delegateFrom(pendingRefs, write));
    }
```

File: ratis-server/src/main/java/org/apache/ratis/server/impl/MessageStreamRequests.java [hunk 2]
```diff
@@ -101,7 +114,7 @@ CompletableFuture<ByteString> streamEndOfRequestAsync(RaftClientRequest request)
     if (pending == null) {
       return JavaUtils.completeExceptionally(new StreamException(name + ": " + key + " not found"));
     }
-    return pending.getBytes(stream.getMessageId(), request.getMessage());
+    return pending.getBytes(stream.getMessageId(), requestRef.delegate(request.getMessage()));
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/MessageStreamRequests.java:117]: Then, we don't have to delegate here.
```java
  CompletableFuture<ReferenceCountedObject<RaftClientRequest>> streamEndOfRequestAsync(
    ...
    return pending.getWriteRequest(stream.getMessageId(), requestRef);
  }
```

File: ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java [hunk 3]
```diff
@@ -534,15 +535,21 @@ PendingRequest addPendingRequest(PendingRequests.Permit permit, RaftClientReques
...
-    return messageStreamRequests.streamEndOfRequestAsync(request)
-        .thenApply(bytes -> RaftClientRequest.toWriteRequest(request, Message.valueOf(bytes)));
+  CompletableFuture<ReferenceCountedObject<RaftClientRequest>> streamEndOfRequestAsync(
+      ReferenceCountedObject<RaftClientRequest> requestRef) {
+    RaftClientRequest request = requestRef.get();
+    return messageStreamRequests.streamEndOfRequestAsync(requestRef)
+        .thenApply(bytesRef -> {
+          RaftClientRequest finalRequest = RaftClientRequest.toWriteRequest(request, Message.valueOf(bytesRef.get()));
+          return bytesRef.delegate(finalRequest);
+        });
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/LeaderStateImpl.java:552]: The code here becomes trivial.
```java
    return messageStreamRequests.streamEndOfRequestAsync(requestRef);
```

File: ratis-server/src/main/java/org/apache/ratis/server/impl/MessageStreamRequests.java [hunk 4]
```diff
@@ -39,25 +42,32 @@ private static class PendingStream {
...
     private ByteString bytes = ByteString.EMPTY;
+    private List<ReferenceCountedObject<Message>> pendingRefs = new LinkedList<>();
 
     PendingStream(ClientInvocationId key) {
       this.key = key;
     }
 
-    synchronized CompletableFuture<ByteString> append(long messageId, Message message) {
+    synchronized CompletableFuture<ReferenceCountedObject<ByteString>> append(long messageId,
+        ReferenceCountedObject<Message> messageRef) {
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/MessageStreamRequests.java:52]: Pass `ReferenceCountedObject<RaftClientRequest> requestRef` instead.

File: ratis-server/src/main/java/org/apache/ratis/server/impl/MessageStreamRequests.java [hunk 5]
```diff
@@ -39,25 +42,32 @@ private static class PendingStream {
     private final ClientInvocationId key;
     private long nextId = -1;
     private ByteString bytes = ByteString.EMPTY;
+    private List<ReferenceCountedObject<Message>> pendingRefs = new LinkedList<>();
```
Comment (szetszwo) [inline ratis-server/src/main/java/org/apache/ratis/server/impl/MessageStreamRequests.java:45]: Store `RaftClientRequest` instead.
```java
    private final List<ReferenceCountedObject<RaftClientRequest>> pendingRefs = new LinkedList<>();
```

**Review by szetszwo — COMMENTED** [review]
@duongkame , thanks for the update!   We need to release `pendingRefs` when the map is being cleared; see the comments inlined and also https://issues.apache.org/jira/secure/attachment/13065937/1007_review.patch

**Comment by duongkame** [comment]
> @duongkame , thanks for the update! We need to release `pendingRefs` when the map is being cleared; see the comments inlined and also https://issues.apache.org/jira/secure/attachment/13065937/1007_review.patch

Thanks a lot for the review @szetszwo.

**Comment by szetszwo** [comment]
@duongkame , thanks a lot for working on this hard problem!

**Comment by szetszwo** [comment]
```
org.apache.ratis.grpc.TestWatchRequestWithGrpc.testWatchRequestAsyncChangeLeader  Time elapsed: 34.557 s  <<< ERROR!
java.util.concurrent.TimeoutException
	at java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1784)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1928)
	at org.apache.ratis.WatchRequestTests.checkMajority(WatchRequestTests.java:274)
	at org.apache.ratis.WatchRequestTests.runTestWatchRequestAsyncChangeLeader(WatchRequestTests.java:360)
	at org.apache.ratis.WatchRequestTests.runTest(WatchRequestTests.java:153)
	at org.apache.ratis.WatchRequestTests.lambda$testWatchRequestAsyncChangeLeader$6(WatchRequestTests.java:337)
	at org.apache.ratis.server.impl.MiniRaftCluster$Factory$Get.runWithNewCluster(MiniRaftCluster.java:140)
	at org.apache.ratis.server.impl.MiniRaftCluster$Factory$Get.runWithNewCluster(MiniRaftCluster.java:120)
	at org.apache.ratis.WatchRequestTests.testWatchRequestAsyncChangeLeader(WatchRequestTests.java:336)
	...
```
`testWatchRequestAsyncChangeLeader` may timeout as above.  In the log, it has the following error.  Will fix it in RATIS-2000.
```
2024-01-18 01:41:51,651 [grpc-default-executor-0] ERROR impl.OrderedAsync (OrderedAsync.java:lambda$sendRequest$9(247)) - client-6F96E8AE8714: Failed* RaftClientRequest:client-6F96E8AE8714->s2@group-4F85081496F8, cid=64, seq=20, RW, m7
java.util.concurrent.CompletionException: java.io.IOException: java.lang.IllegalStateException: Failed to retain: object has already been completely released.
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:292)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:308)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:607)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:591)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
	at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1990)
	at org.apache.ratis.grpc.client.GrpcClientProtocolClient$AsyncStreamObservers.completeReplyExceptionally(GrpcClientProtocolClient.java:394)
	at org.apache.ratis.grpc.client.GrpcClientProtocolClient$AsyncStreamObservers.access$000(GrpcClientProtocolClient.java:300)
	at org.apache.ratis.grpc.client.GrpcClientProtocolClient$AsyncStreamObservers$1.onError(GrpcClientProtocolClient.java:331)
	at org.apache.ratis.thirdparty.io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:481)
	at org.apache.ratis.thirdparty.io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at org.apache.ratis.thirdparty.io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at org.apache.ratis.grpc.metrics.intercept.client.MetricClientCallListener.onClose(MetricClientCallListener.java:47)
	at org.apache.ratis.thirdparty.io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:567)
	at org.apache.ratis.thirdparty.io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:71)
	at org.apache.ratis.thirdparty.io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:735)
	at org.apache.ratis.thirdparty.io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:716)
	at org.apache.ratis.thirdparty.io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at org.apache.ratis.thirdparty.io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
Caused by: java.io.IOException: java.lang.IllegalStateException: Failed to retain: object has already been completely released.
	at org.apache.ratis.grpc.GrpcUtil.unwrapException(GrpcUtil.java:103)
	at org.apache.ratis.grpc.GrpcUtil.unwrapIOException(GrpcUtil.java:163)
	at org.apache.ratis.grpc.client.GrpcClientProtocolClient$AsyncStreamObservers$1.onError(GrpcClientProtocolClient.java:330)
	... 13 more
Caused by: java.lang.IllegalStateException: Failed to retain: object has already been completely released.
	at org.apache.ratis.util.ReferenceCountedObject$4.retain(ReferenceCountedObject.java:183)
	at org.apache.ratis.server.impl.RaftServerProxy.submitClientRequestAsync(RaftServerProxy.java:450)
	at org.apache.ratis.grpc.server.GrpcClientProtocolService$RequestStreamObserver.processClientRequest(GrpcClientProtocolService.java:248)
	at org.apache.ratis.grpc.server.GrpcClientProtocolService$OrderedRequestStreamObserver.processClientRequest(GrpcClientProtocolService.java:364)
	at org.apache.ratis.util.SlidingWindow$Server.processRequestsFromHead(SlidingWindow.java:459)
	at org.apache.ratis.util.SlidingWindow$Server.receivedRequest(SlidingWindow.java:451)
	at org.apache.ratis.grpc.server.GrpcClientProtocolService$OrderedRequestStreamObserver.processClientRequest(GrpcClientProtocolService.java:389)
	at org.apache.ratis.grpc.server.GrpcClientProtocolService$RequestStreamObserver.onNext(GrpcClientProtocolService.java:272)
	at org.apache.ratis.grpc.server.GrpcClientProtocolService$RequestStreamObserver.onNext(GrpcClientProtocolService.java:191)
	at org.apache.ratis.thirdparty.io.grpc.stub.ServerCalls$StreamingServerCallHandler$StreamingServerCallListener.onMessage(ServerCalls.java:262)
	at org.apache.ratis.thirdparty.io.grpc.ForwardingServerCallListener.onMessage(ForwardingServerCallListener.java:33)
	at org.apache.ratis.thirdparty.io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.messagesAvailableInternal(ServerCallImpl.java:329)
	at org.apache.ratis.thirdparty.io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.messagesAvailable(ServerCallImpl.java:314)
	at org.apache.ratis.thirdparty.io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1MessagesAvailable.runInContext(ServerImpl.java:833)
	... 5 more
```

**Pull request merged**